src/components/AttachmentCarousel/AttachmentCarouselView/index.js

```diff
@@ -0,0 +1,152 @@
+import React, {useRef, useCallback} from 'react';
+import {View, FlatList, PixelRatio} from 'react-native';
+import PropTypes from 'prop-types';
+import styles from '../../../styles/styles';
+import CarouselActions from '../CarouselActions';
+import AttachmentView from '../../AttachmentView';
+import withWindowDimensions, {windowDimensionsPropTypes} from '../../withWindowDimensions';
+import CarouselButtons from '../CarouselButtons';
+
+const propTypes = {
+    // eslint-disable-next-line react/forbid-prop-types
+    carouselState: PropTypes.object.isRequired,
+
+    updatePage: PropTypes.func.isRequired,
+    toggleArrowsVisibility: PropTypes.func.isRequired,
+    autoHideArrow: PropTypes.func.isRequired,
+    cancelAutoHideArrow: PropTypes.func.isRequired,
+
+    ...windowDimensionsPropTypes,
+};
+
+function AttachmentCarouselView(props) {
+    const scrollRef = useRef(null);
+    const viewabilityConfig = {
+        // To facilitate paging through the attachments, we want to consider an item "viewable" when it is
+        // more than 90% visible. When that happens we update the page index in the state.
+        itemVisiblePercentThreshold: 95,
+    };
+
+    /**
+     * Calculate items layout information to optimize scrolling performance
+     * @param {*} data
+     * @param {Number} index
+     * @returns {{offset: Number, length: Number, index: Number}}
+     */
+    const getItemLayout = useCallback(
+        (data, index) => ({
+            length: props.carouselState.containerWidth,
+            offset: props.carouselState.containerWidth * index,
+            index,
+        }),
+        [props.carouselState.containerWidth],
+    );
```

Comment:
This question is purely to help me understand hooks, so thanks for your patience :)

I'd love to know why useCallback is preferred to useMemo here? Is it because it is very unlikely that the `containerWidth` prop will ever be the same number twice? Also because it could change 100's of times as the user changes width? (perhaps the callback happens AFTER resize, but for now I assume it is triggered on each width int change)

Reply:
From my understanding, `getItemLayout` needs to be a function rather than the actual "`layout`" value directly. `FlatList`/`VirtualizedList` calls this function itself, it's not just using the value